src/main/dynamo/Beyonce.ts

```diff
@@ -149,12 +152,13 @@ export class Beyonce {
 
     if (responses !== undefined) {
       const items = responses[this.table.tableName]
-      const jsonItemPromises = items.map(async (_) => {
-        const item = await decryptOrPassThroughItem(this.jayz, _)
-        return toJSON<ExtractKeyType<T>>(item)
-      })
-
-      const jsonItems = await Promise.all(jsonItemPromises)
+      const maybeDecryptedItems = await decryptOrPassThroughItems(
```

Comment:
Note how we generate fewer promises here now (1 instead of N), which should translate into less overhead.

Reply:
Nice.

Under the hood, do we do something like this?   (scala pseduo-code below)

```
// for encryption
items.grouped(batchSize).map { batch =>
  // encrypt all batch items with same encryption key, ideally on a worker thread
}

// for decryption
items.groupBy(decryptionKeyId).map { sameKeyBatch =>
  // fetch decryption key
  batch.grouped(decryptionBatch).map {
    // decrypt batch with same decryption key on worker thread
  }
}
```

Reply:
Well...sort of. 

It's sort of similar in that:  we will batch encrypt groups of things using the same key. But, we'll swap out the key in the loop as needed whenever we hit `maxUsesPerDataKey`, which you can [see here](https://github.com/ginger-io/jay-z/blob/master/src/main/JayZ.ts#L104). 

But it's pretty different in that:

1. The Jay-Z encryption library, which is what is getting called under the covers here, is currently single threaded. Node does actually have worker threads now. And, they're available in 12.x (what we run on AWS) -- so it'd be a nice future upgrade. Unfortunately, the [API isn't spectacular](https://nodejs.org/docs/latest-v12.x/api/worker_threads.html#worker_threads_worker_threads) and involves loading a separate script/file for your worker.

2. When you encrypt stuff with a KMS data key it appends metadata to the encrypted byte that indicate what KMS key id was used to decrypt it. So, when we decrypt we're very naive and aren't caching keys we've already retrieved. 

My mental roadmap for our Jay-Z library looks something like: 

1. Append the KMS key id to the encrypted metadata in a way that we can read it + batch items to decrypt that need the same key so we only make 1 network request per key. 

2. Parallelize encryption/decryption + maybe default to 1 worker thread per cpu.

Reply:
... Also, I think the typical encryption batch size is going to be fairly small < 100 objects, so would be interesting to benchmark to see if using worker threads for encryption/decryption actually helps or hinders there.

Reply:
Sounds good, agreed on roadmap.   My sense is that if we do parallelize with workers, it will change the structure of the code because we'll want to avoid concurrent access to the same data (e.g. cached keys).   Workers/isolates are share-nothing right?

Also agreed that parallelization will mostly help (reduce latency) in the cases of large batches.  Another approach would be to push parallelism at a higher level, to be managed by the client if need be.

Reply:
Agreed on the structure of the code would look different if parallelized. 

re: (network) latency -- hmm I'm not sure that adding worker threads would reduce that further here. Right now, we should be doing non-blocking IO already in that we fire N network requests concurrently and only await their results as a group (vs sequentially/individually). But, it seems like a good thing to try benchmarking using a worker thread vs not. 

re: where to put parallelism ... yea if we add it, hoisting it up another level seems like a good way to go as it'd allow for parallelizing other cpu intensive things (e.g. json parsing + post processing of DB results).

> Workers/isolates are share-nothing right?

Actually, you can choose between: 

1. Shared memory accessible to both the main and worker thread via [SharedArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer) 

2. Transfer (copy) the memory to the worker thread via  [postMessage](https://nodejs.org/docs/latest-v12.x/api/worker_threads.html#worker_threads_port_postmessage_value_transferlist)'s `transferList`.

3. Just copy the memory (default thing that happens).

Reply:
Thanks for the pointers, and good point about parallelizing other things while we're at it.

Reply:
No prob, thanks for the feedback and discussion!